### ISSUE TITLE
Possible I/O performance optimization for the RPMB filesystem

### ISSUE THREAD

**Issue by ms-mahuber** [OP-TEE/optee_os#3641]
I would like to discuss a possibility for I/O performance optimization for the RPMB filesystem (fs).
Part of the fs implementation is the function 'read_fat'.
'read_fat' is frequently called when interacting a lot with RPMB storage and can have a potentially high overhead.
This is due to 'read_fat' traversing the fat fs and reading in fat fs entries ('struct rpmb_fat_entry') through calling the function 'tee_rpmb_read'.
Calling 'tee_rpmb_read' results in I/O through the ns-world driver.
A previous optimization to read in a higher number of fat fs entries with a single 'tee_rpmb_read' call can reduce the overhead of 'read_fat' (https://github.com/OP-TEE/optee_os/pull/3607).
Nevertheless, as 'read_fat' may be a frequently called function, 'read_fat' must at least once make a call to 'tee_rpmb_read'.

An idea for further optimization when frequently interacting with RPMB storage could be the following:
Keep a "shadow list" (alias "cache") of fat fs entries in secure world main memory.
Doing so, 'read_fat' could read from this "shadow list" instead of calling 'tee_rpmb_read' to read from RPMB.
Whenever we modify a FAT entry through calling 'write_fat_entry', the "shadow list" in main memory has to be updated.
One could make this feature an opt-in feature for platforms with sufficient main memory as the "shadow list" can potentially consume a larger amount of secure world main memory.

What do you think about the reasonability of this idea and its feasibility?
Would this be a change that is solicited/appreciated?

An open question on my side is the interplay of the 'tee_mm_pool_t *p' parameter within 'read_fat'.
I haven't fully grasped how that memory pool concept works exactly.
Possibly, that could be used for the construction of such a "shadow list".

**Comment by jbech-linaro** [comment]
Sounds like a solution that could work, I was first worried about the writing part, but it seems like you've thought about that. However, note that a data read in RPMB is initiated by a Write Multiple Block command, I'm not sure if and how that would affect something like this.

**Comment by ms-mahuber** [comment]
Hi, thanks for your reply. As of now, I see no conflict with 'that a data read in RPMB is initiated by a Write Multiple Block command'. First of all, keeping an in-RAM data structure that contains FAT fs entries could reduce the amount of RPMB reads (tee_rpmb_read) required through the lookups of FAT fs entries that 'read_fat' does. Second, any write 'multiple block command' in the REE driver (such as a UEFI or OS driver) caused by a data read should (must?) not affect information stored in FAT fs entries (see 'struct rpmb_fat_entry') - according to my understanding. These FAT fs entries in RPMB storage are solely updated through calling 'write_fat_entry'.

Please let me know if I am missing something.

In the meanwhile, I'd be able to allocate some time to come up with a more concrete sketch of how an implementation could look like (and to investigate its feasibility/complexity/benefit).

**Comment by jbech-linaro** [comment]
Right, I think we're on the same page, but for a change like this one of the challenges in implementing a cache, is where to cut the connection with ordinary RPMB calls, but I get the impression that you've have a somewhat clear picture of it. Thanks!

**Comment by ms-mahuber** [comment]
I sketched an implementation that I will be testing next week. The cache/buffer caches the first n FAT fs entries and can further reduce the amount of RPMB reads OPTEE needs to make. To make sure I'm not working towards the wrong direction, I wanted to provide you with a high-level overview of the changes to core/tee/tee_rpmb_fs.c I envisage (let's keep naming conventions etc. out of scope for now and just consider the general idea):
- added a data structure that represents the buffer/cache for the FAT fs entries
- added a static instance of that data structure
- added functions that initialize, de-initialize, update the buffer/cache, and get the next entry:
    - static TEE_Result fat_entry_dir_init(void)
    - static void fat_entry_dir_deinit(void)
    - static TEE_Result fat_entry_dir_update(struct rpmb_fat_entry *fat_entry, uint32_t fat_address)
    - static TEE_Result fat_entry_dir_get_next(struct rpmb_fat_entry **fat_entry, uint32_t *fat_address)

The data structures and functions are part of the code regardless whether caching is on or not. The added functions simplify the way we make FAT fs traversals. We currently make these traversals from within three functions:
- read_fat
- dump_fat
- rpmb_fs_dir_populate

In the three functions we now basically call the following functions to conduct a FAT fs traversal:
1. fat_entry_dir_init
2. loop with fat_entry_dir_get_next
3. fat_entry_dir_deinit

Regardless whether caching is on or not, these functions encapsulate the RPMB reads and buffering/caching of FAT fs entries, which simplifies read_fat, dump_fat, rpmb_fs_dir_populate (and any imaginary future function doing a FAT fs traversal).

Whenever we write a FAT fs entry with write_fat_entry, we call the new function fat_entry_dir_update after the FAT fs entry has been persisted. fat_entry_dir_update updates the respective element in the cache if caching is enabled.

I know that the text above provides only limited insight, but before stepping to quickly ahead, I wanted to share more of my idea to hear what you think about it.

**Comment by jenswi-linaro** [comment]
Makes sense to me.

**Comment by ms-mahuber** [comment]
I made an implementation and tested it with different cache sizes etc. On the hardware-software configuration I am working on, my patch works. I added quite some documentation to the code. As a next step, I would make a pull request (by around tomorrow). ...but I'm also happy to discuss/present more details in this thread before making the pull request.

An example configuration could be:
- number of FAT FS entries to be read in = 8 (existing config variable)
- cache size = 16
- Reading FAT FS entries, this would mean: When the first FAT FS entry is requested from RPMB, we read in 8 entries and store them in the cache. Whenever we read beyond the last cached element, we fetch in the next 8 elements. With this, the cache is full.
- If there is more than 16 FAT FS entries that we need to read, we allocate a temporary ('scratchboard') buffer that fetches the next 8 entries. If we need further entries, this buffer is overwritten with the subsequent 8 entries, etc. When a FAT traversal completes, only the cached elements are retained.
- With caching disabled, we only use this temporary buffer - this results in exactly the same behavior as in the previous implementation. This means, with caching enabled we require (at most) 'cache size' * sizeof(struct rpmb_fat_entry) additional heap memory (e.g. 16*256 byte). On platforms where that memory is available, caching can significantly improve RPMB I/O.

**Comment by jenswi-linaro** [comment]
This looks very interesting. :-)

**Comment by ms-mahuber** [comment]
I've created a PR: https://github.com/OP-TEE/optee_os/pull/3672

Please, feel free to leave any comments. All naming/conventions etc. can be discussed, I'm really happy to learn. A good point for more documentation is the mk/config.mk

I've tested the patch on my platform with various options (caching off, caching on, different cache size, read in sizes), but apart from possible coding improvements, it would be great if it could be tested on other platforms.

**Comment by ms-mahuber** [comment]
I have several 'check' flags in the Travis CI. There is no warnings and errors, but it's very difficult for me to get rid of the 'check' statements. I think this is ultimately what the Travis CI pipeline complains about.

**Comment by jforissier** [comment]
> I have several 'check' flags in the Travis CI. There is no warnings and errors, but it's very difficult for me to get rid of the 'check' statements.

On the contrary, all of them seem quite easy to fix to me ;-) You may use temporary variables with shorter names for instance, for lines which would run too far.

**Comment by ms-mahuber** [comment]
I've updated the patch, thanks for the tips. You were right, of course. I think, the changes make the patch easier to conceive.

One of the major aspects for this patch would be to do some more RPMB testing than only on my platform. I guess, BART has no testing for the RPMB FS (which would require emulation etc. for physical RPMB...)? It would be great if we could identify somebody that actually makes use of the RPMB FS and test the patch on their platform.

**Comment by jforissier** [comment]
FYI, IBART does test with RPMB emulation enabled on the HiKey 620 board, see here: https://github.com/jbech-linaro/ibart/blob/master/jobdefs/examples/hikey6220.yaml#L38

**Comment by ms-mahuber** [comment]
Hi Jérôme,

thanks for your reply and for kicking off this discussion. :)
First of all, be assured that we have thought about implementing an LRU cache.
However, our conscious choice was to withdraw this idea.
In the following I'll try to object to your concern regarding 1) 'no speed benefit', 'performance is likely to degrade abruptly' and 2) regarding the idea of implementing an LRU cache.
3) Lastly, I'll provide a summary.

1) Performance of proposed solution (cache first N) compared to current solution (no cache)

Some variables to introduce for the remainder:
num_entries_total = Number of entries in FAT FS
num_entries_cache = Number of entries FAT FS cache can hold
num_read_in = Number of entries read into the buffer/cache with one invocation of rpmb_read

- Current solution:

Everytime we read entries from the FAT FS, we have to make at least one invocation of rpmb_read to fetch FAT FS entries.
In the worst case, we have to read 'ceil(num_entries_total/num_read_in)' times to find an entry (e.g. num_entries_total=44, num_read_in=10, the worst case is 5 reads).
At least with our use case, this creates a lot, lot of rpmb_read calls.

This is why we were seeking for an optimization, resulting in our proposed solution.

- Proposed solution:

In the best case, our cache fits all FAT FS entries, i.e., num_entries_total <= num_entries_cache.
In this case we need to only fill the cache, and then no longer have to deal with any read-ins anymore.
The number of read-ins to fill the cache would be ceil(num_entries_total/num_read_in).
Notably, this is equal to the worst case number of reads for read a single FAT FS entry with no cache, so we make big savings.
We are in the best case as long as num_entries_total <= num_entries_cache.
In a lot of use cases this number can be foreseen at build time.

Let's discuss the case where num_entries_total > num_entries_cache, i.e., where the cache is too small to hold all elements.
Let's consider the worst case where we only access elements that cannot be cached ('not the first N elements in the FAT FS').
Thanks to the cache, we know we can skip the first num_entries_cache entries and start reading the FAT FS from the first entry that is not in the cache.
In this case, we do always better than the current solution, because we require less read-ins.
We use a temporary buffer to read in the entries that do not fit into the cache, and read from the FAT FS until we find the entry.
The worst case of read-ins required for finding a FAT FS entry in this case is ceil((num_entries_total-num_entries_cache)/num_read_in), where 'num_entries_total-num_entries_cache' is 'the number of entries not fitting into the cache'.

Sidenote: More explanation on the cache documented in the code and in the related issue (https://github.com/OP-TEE/optee_os/issues/3641)


2) Having the points discussed in 1) in mind, let's consider an LRU cache

The first thing I'd like to point out is that in this scenario we should have an initial non-lazy LRU cache-fill implementation.
If the cache is lazy, the LRU implementation is likely to perform bad.
Imagine we read the first entry, so we need to traverse the FAT FS until we find it and add it to the cache.
Reading the next entry, we need to traverse the FAT FS again, because the next entry is most likely not cached.
Note that our proposed solution works in a lazy manner.
We incrementally fill the cache by the size 'num_read_in' until we read in 'num_entries_cache' entries into the cache.
Of course, that non-lazy implementation for the LRU is not a big downside.
We can also add entries that we read in until finding the proper entry to the cache and thereby populate the cache.
Or, we'd have an initial read-in phase where the cache is populated.
But it's a particularity that needs to be kept in mind, and I don't think this aspect is favorable compared to our implementation.

Now, let's consider the best case.
We fill the LRU cache with num_entries_total, as num_entries_total <= num_entries_cache.
In this case, the cache should perform similar to our proposed solution.

Let's consider the worst case, or better, a case where we frequently need to operate on elements that are evicted from the LRU cache.
The probability that this happens depends on num_entries_total vs num_entries_cache.
In this case, we have to read in from the FAT FS and at this point, it gets complicated.
The question is how do we read in.
We most likely have to read from the start of the FAT FS entries (unless we hold additional metadata but then the implementation becomes even more complex).
In the case we have to read entries in, I do not see an upside compared to our current solution.
As in our solution, we need a temporary buffer and read in 'num_read_in' entries to determine whether our entry is contained.
In the worst case, this results in a full FAT traversal.
This case does never happen with our proposed solution, and could lead to the performance degradation you were hesitant about.
An example for this is X=cached, .=non-cached, FATFS from entry 1...n is [X...XX..X......X.XXX...]
Without holding further metadata and complicating an implementation, we'd most likely have to read from the start.

Having said this, I don't want to argue that our proposed solution is better and I did not make a science to determine an average case scenario or probabilities for the worst case scenario for both solutions. ;)
Most of all, not indisposed to better future solutions - but confident that our proposed solution is beneficious.

3) Summary: 
In a lot of use cases the number of expected FAT FS entries can be foreseen/estimated at build time and the cache adjusted accordingly, making our solution a good choice.
Further, on a lot of platforms we have sufficient RAM for a cache to hold a substantial amount of entries, or all entries.
And, we proposed our cache as an optional feature.
The cache works for our use case very well and yields significant performance gain.
We have a deployment where we fit all entries into the cache, but also tested with a cache that is too small, and also this scenario gives us performance gain.
For use cases, where the cache fits only a small fraction of elements, and/or where the number of elements cannot be foresee/estimated, we have argued that the cache does improves performance and does not degrade it.
If there are any remaining concerns regarding the that the cache would not improve the 'current solution', a guidance can be that the cache should be big enough when making use of this feature, the argumentation above should be sane.
Further, we chose this implementation for its simplicity.
We'd argue that this cache implementation is at least a step in the right direction towards RPMB performance and that this idea and implementation can of course be leveraged for further optimizations when it is streamlined.

Really curious about your opinion.
Thanks for the discussion and most of all for reading if you came until down here,

Manuel
